driver-core/src/main/com/mongodb/operation/MixedBulkWriteOperation.java

```diff
@@ -265,6 +265,9 @@ private BulkWriteResult executeBulkWriteBatch(final WriteBinding binding, final
                         throw new MongoWriteConcernWithResponseException(writeConcernBasedError, result);
                     }
                 }
+                if (binding.getSessionContext().hasActiveTransaction()) {
```

Comment:
It's not sufficient to only set the recovery token here.  Consider for example a transaction that uses findAndModify, or aggregate with $out.  

The good thing is that there is already a place where it makes sense to slot this in: all the way down in `InternalStreamConnection#updateSessionContext`.